### ISSUE TITLE
Token issuing of sponsorship renewal

### ISSUE THREAD

**Issue by jrosindell** [OneZoom/OZtree#393]
We need a system to issue the tokens by e-mail message to enable renewal of sponsorships.

This page should be a very simple one that allows people to enter their e-mail address and get e-mailed a token to use for renewal.

Some kind of debug mode when enabled should permit us when testing to get the tokens not via e-mail.

**Comment by lentinj** [comment]
Presumably we ignore ``restrict_all_contact`` if they've filled in the form manually.

Do we error on an e-mail address without any reservations? Or fail silently? Latter is probably better, if we don't mind any potential debugging confusion.

> Some kind of debug mode when enabled should permit us when testing to get the tokens not via e-mail.

Would "If SMTP settings aren't set, we display the URL in the flash message" suffice? Or would you still want to do this on the live server?

**Comment by hyanwong** [comment]
Yes, the T&C explicitly allow for us to contact them for renewal, even if they have selected "restrict all contact". There's a field "emailed_re_renewal_initial" which we should fill out when the first email is sent.

**Comment by lentinj** [comment]
Right, the above is a first-stab at this:

* Currently it explicitly tells users if it can't find a matching reservations entry, we might want to hide that message
  * This would also be where we do a e-mail --> username lookup, if we can't lookup it means we don't have one. We then email a signed link to ``sponsor_renew.html/UserName``.
* The HMAC keys for the signature are UUIDs that get cached and rotated, this means links will expire when the keys do, but offering permanently-valid links means that we'd have to remember to set a key in ``appconfig.ini``, and if it gets discovered then all links get invalidated when changing it. This seemed a better approach.
* If you don't have SMTP set up in ``appconfig.ini``, then you get flash-messaged the link.

**Comment by hyanwong** [comment]
Thanks @lentinj - where are the UUIDs stored then, and how to they expire?

**Comment by lentinj** [comment]
> where are the UUIDs stored then,

In the disk cache, e.g. ``OZtree/cache/1c0/195/sponsor_renew_hmac_keys``. It uses the disk cache over the ram cache specifically because of restarts.

>  and how to they expire?

On link generation, if they're too old then then the oldest of the pair gets the boot and is regenerated. There's a pair so you don't get caught in the cutoff.

**Comment by jrosindell** [comment]
A couple of thoughts on this.....

First, I think that it should fail silently if the e-mail isn't in the system - this way people can't extract information about whether e-mails are registered or not.

Second, what about the page in #364 these links should be permanent right and therefore not using these tokens, but they should also not be using the e-mail address so how will they be done?  I don't think it's good if people's personal pages have temporary URLs

**Comment by hyanwong** [comment]
> First, I think that it should fail silently if the e-mail isn't in the system - this way people can't extract information about whether e-mails are registered or not.

Yes, I agree

> Second, what about the page in #364 these links should be permanent right and therefore not using these tokens, but they should also not be using the e-mail address so how will they be done? I don't think it's good if people's personal pages have temporary URLs

They don't use the token, but do have the `reservations.username`. So the link will be onezoom.org/donor/YanWong, I guess.

**Comment by hyanwong** [comment]
Is there any reason we shouldn't mock up a permanent private URL using a *really long* HMAC renewal time? E.g. 4 years? Or perhaps store the HMAC in the database? I think I'm not quite understanding how it all works here.

**Comment by lentinj** [comment]
> using a really long HMAC renewal time

Yup, we can just bump the expiry. We still have to have 2, in case you try and renew the day before the switchover

> Or perhaps store the HMAC in the database

Also a reasonable option, but seemed like more development than it was worth, given the ``cache`` modules are already available.

**Comment by lentinj** [comment]
> They don't use the token, but do have the reservations.username. So the link will be onezoom.org/donor/YanWong, I guess. See #394

That's what's currently happening with https://github.com/OneZoom/OZtree/commit/90ee68b6b16c681137b0a190c61866a33b5f52c3

**Comment by lentinj** [comment]
> Since we are changing the DB anyway, do we want to add an (unused) "HMAC" field to reservations

The HMAC is global (well, per-server), not per-user. It's the private key that means we know this URL was generated by the server.

If we had a per-user column, we could just give a user a UUID and then use that as the URL string.

**Comment by hyanwong** [comment]
> > Since we are changing the DB anyway, do we want to add an (unused) "HMAC" field to reservations
> 
> The HMAC is global (well, per-server), not per-user. It's the private key that means we know this URL was generated by the server.

Ah, I see. Sorry. I was being stupid.

> If we had a per-user column, we could just give a user a UUID and then use that as the URL string.

Yep, ISWYM. 

The only thing I'd be worried about would be the storing of server state somewhere else - yet another thing to remember when moving the site around, etc. At the moment it's nice that everything site-specific is either in the DB or in appconfig.ini. I guess we could just put a comment in the appconfig.ini file that the cached HMAC keys need preserving too, if the site is to be replicated identically elsewhere.

**Comment by lentinj** [comment]
The theory was that using a pre-generated cached value would make your life easier, so you don't have to think about it. If you'd rather an ``appconfig.ini`` value with a secret then happy to switch to that.

I think I'd rather error than set a random default value (for test setups), so it's obvious you need to for a production setup.

**Comment by hyanwong** [comment]
On reflection I think I prefer a secret appconfig.ini value, yes.

**Comment by hyanwong** [comment]
Great, thanks.

**Comment by jrosindell** [comment]
I'm guessing that this is actually quite done and that Yan and I need to test?

**Comment by lentinj** [comment]
@jrosindell I believe so. The renewals request page might need linking up in places, but I'm guessing users triggering the e-mail manually isn't the most likely way it gets used.
